src/internal/storage/fileset/index/doc.go

```diff
@@ -0,0 +1,19 @@
+/*
+Package index implements a hierarchical file system index.
+
+Each element in the hierarchy is an [index.Index], which refers to data in object storage via a [chunk.DataRef].
+The referenced data is either the contents of a single file or a list of other indices spanning a given path range, stored as a marshalled protos.
+A hierarchy always has a single top-level index, which stored in e.g. a file set's additive field to indicate its contents.
+
+The [index.Reader] and [index.Writer] automatically traverse intermediate index layers, allowing consumers to only interact with file indices.
+The reader makes use of the hierarchical range representation to allow quickly skipping to sub-ranges, directories, or individual files.
```

Comment:
I would just say subranges. That captures individual files and directories are not a concept at this level.